### PR TITLE
SDKS-2631 Android App Integrity Implementation

### PR DESCRIPTION
**Pull request by witrisna**
# JIRA Ticket

[SDKS-2631](https://bugster.forgerock.org/jira/browse/SDKS-2631)

# Description

SDKS-2631 Android App Integrity Implementation

### CODE REVIEW

File: forgerock-auth/src/main/java/org/forgerock/android/auth/callback/AppIntegrityCallback.kt [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+                RequestType.CLASSIC -> {
+                    val integrityManager = getIntegrityManager(context)
+                    val builder = IntegrityTokenRequest.builder()
+                    builder.setCloudProjectNumber(projectNumber.toLong())
+                    builder.setNonce(hashWithNonce(getAuthId()))
+                    setToken(integrityManager.requestIntegrityToken(builder.build()).await()
+                        .token())
+                }
+
+                /*
```
Comment (jeyanthanperiyasamy) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/callback/AppIntegrityCallback.kt:129]: is this for future implementation ?

**Comment by jeyanthanperiyasamy** [comment]
looks good , pretty straightforward

**Review by jeyanthanperiyasamy — APPROVED** [review]

File: forgerock-auth/src/main/java/org/forgerock/android/auth/callback/AppIntegrityCallback.kt (hunk 1 above)
Comment (witrisna) [reply forgerock-auth/src/main/java/org/forgerock/android/auth/callback/AppIntegrityCallback.kt:129]: Yes, this is new from Google, would like to get it ready. I have high-level tested it, and it works as expected, but not ready to make it official at this moment.

File: forgerock-auth/src/main/java/org/forgerock/android/auth/Node.java [hunk 2]
```diff
@@ -38,6 +33,15 @@ public class Node implements Serializable {
     private final String authServiceId;
     private final List<Callback> callbacks;
 
+    public Node(String authId, String stage, String header, String description, String authServiceId, List<Callback> callbacks) {
```
Comment (rodrigoareis) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/Node.java:36]: Javadoc?

File: forgerock-auth/src/main/java/org/forgerock/android/auth/Node.java [hunk 3]
```diff
@@ -123,4 +126,23 @@ public void setCallback(Callback callback) {
         }
     }
 
+    public String getAuthId() {
```
Comment (rodrigoareis) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/Node.java:129]: Javadoc?

File: forgerock-auth/src/main/java/org/forgerock/android/auth/callback/CallbackFactory.java [hunk 4]
```diff
@@ -79,4 +77,7 @@ public String getType(Class<? extends Callback> callback) throws InstantiationEx
         return callback.newInstance().getType();
     }
 
+    public Map<String, Class<? extends Callback>> getCallbacks() {
```
Comment (rodrigoareis) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/callback/CallbackFactory.java:80]: Javadoc?

**Review by rodrigoareis — COMMENTED** [review]
Overall implementation looks good. Minor comments regarding javadoc

File: forgerock-auth/src/main/java/org/forgerock/android/auth/Node.java (hunk 3 above)
Comment (witrisna) [reply forgerock-auth/src/main/java/org/forgerock/android/auth/Node.java:142]: javadoc updated.

File: forgerock-auth/src/main/java/org/forgerock/android/auth/callback/CallbackFactory.java (hunk 4 above)
Comment (witrisna) [reply forgerock-auth/src/main/java/org/forgerock/android/auth/callback/CallbackFactory.java:80]: javadoc updated.

File: forgerock-auth/src/main/java/org/forgerock/android/auth/Node.java (hunk 2 above)
Comment (witrisna) [reply forgerock-auth/src/main/java/org/forgerock/android/auth/Node.java:39]: Mark is as internal use only, does not expect developer to create the Node.

**Review by rodrigoareis — DISMISSED** [review]
LGTM

**Review by spetrov — APPROVED** [review]
Looks good! 👍🏻

**Pull request merged**